arbnode/sync_monitor.go

```diff
@@ -0,0 +1,128 @@
+package arbnode
+
+import (
+	"sync/atomic"
+
+	"github.com/offchainlabs/nitro/arbutil"
+	flag "github.com/spf13/pflag"
+)
+
+type SyncMonitor struct {
+	config      *SyncMonitorConfig
+	inboxReader *InboxReader
+	txStreamer  *TransactionStreamer
+	coordinator *SeqCoordinator
+	initialized bool
+}
+
+func NewSyncMonitor(config *SyncMonitorConfig) *SyncMonitor {
+	return &SyncMonitor{
+		config: config,
+	}
+}
+
+type SyncMonitorConfig struct {
+	BlockBuildLag     uint64 `koanf:"block-build-lag"`
+	CoordinatorMsgLag uint64 `koanf:"coordinator-msg-lag"`
+}
+
+var DefaultSyncMonitorConfig = SyncMonitorConfig{
+	BlockBuildLag:     20,
+	CoordinatorMsgLag: 200,
+}
+
+func SyncMonitorConfigAddOptions(prefix string, f *flag.FlagSet) {
+	f.Uint64(prefix+".block-build-lag", DefaultSyncMonitorConfig.BlockBuildLag, "allowed lag between messages read and blocks built")
+	f.Uint64(prefix+".coordinator-msg-lag", DefaultSyncMonitorConfig.CoordinatorMsgLag, "allowed lag between local and remote messages")
+}
+
+func (s *SyncMonitor) Initialize(inboxReader *InboxReader, txStreamer *TransactionStreamer, coordinator *SeqCoordinator) {
+	s.inboxReader = inboxReader
+	s.txStreamer = txStreamer
+	s.coordinator = coordinator
+	s.initialized = true
+}
+
+func (s *SyncMonitor) SyncProgressMap() map[string]interface{} {
```

Comment:
The data checked here is taken from either local database or redis. These are all relatively short read-only calls. Currently all data is retrieved on any call to Synced or eth_syncing.
Two other options:
* cache results for a limited/configurable time. Retrieve new result if cache is old.
* have a background thread constantly polling data, this function always uses cache and returns immediately.
Advantage of having this function always retrieve data is that it's a little more up-to-date and a little better at detecting stuck threads.
What do you think?

Reply:
As discussed, I think it's fine without caching